### PR TITLE
feat: Typesense integration

### PR DESCRIPTION
**Pull request by jasonbosco**
## What

Following up from my conversation with @olivermrbl about collaborating on a [Typesense](https://typesense.org) <> Medusa integration, this is a rough first *draft* of what an integration could look like. 

I've left a couple of questions as TODOs that I'd love to get help on.

## Why

A Typesense user asked for an integration with Medusa.

## How

I've essentially used the skeleton that Oliver put together and filled in Typesense-specific API calls. 

## Testing

Since I'm not too familiar with Medusa, I would like some help from the Medusa team to help test and take this over the finish line. Happy to help with Typesense-related things.

### CODE REVIEW

**Review by olivermrbl — COMMENTED** [review]
This is gonna be so cool @jasonbosco - great work! 🚀 

Have added a couple of comments here and there.

And if you wanna test out the plugin against a running Typesense instance, you should be able to get started by following these steps:

- [Create a Medusa server](https://github.com/medusajs/medusa#-quickstart) or use `npx`
- Navigate to your plugin package, `medusa-plugin-typesense`, run `yarn link`
- Build the plugin `yarn build`
- Navigate to your Medusa server and run `yarn`
- Then run `yarn link medusa-plugin-typesense`
- Add the plugin with its options to `medusa-config.js`
e.g. 
```
{
    resolve: `medusa-plugin-typesense`,
    options: {
    ....
    },
  },
```
- Restart your server -> this should trigger an indexing of products
- Query the Typesense search plugin
```
curl -X POST http://localhost:9000/store/products/search \
  --header "Content-Type: application/json" \
  --data '{"q":"t-shirt"}'
```

Let me know, if the above steps needs elaboration or you can't make it work! I'm ready to jump on a call and assist 🤟

File: packages/medusa-plugin-typesense/src/services/typesense.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  // TODO: Can we access "options" here?
+  async deleteAllDocuments(indexName, options) {
+    await this.client_.collections(indexName).delete();
+    return this.createIndex(indexName, options)
+  }
+
+  search(indexName, query, options) {
+    // TODO: What is the format of `filter` below?
```
[the remark this answers is not in the document]
Comment (olivermrbl) [reply packages/medusa-plugin-typesense/src/services/typesense.js:49]: Looping in @zakariaelas (lead on our Search API) to verify, that what I am saying is not completely wrong 😅  Let us know Zak!

[the remark this answers is not in the document]
Comment (olivermrbl) [reply packages/medusa-plugin-typesense/src/services/typesense.js:49]: Looping in @zakariaelas (lead on our Search API) to ensure, that I am not misguiding you here 😅  Let us know Zak!

[the remark this answers is not in the document]
Comment (zakariaelas) [reply packages/medusa-plugin-typesense/src/services/typesense.js:49]: Correct! As @olivermrbl explained, the `filter` param does not follow a specific structure and is vendor-specific. Took a look at what you currently have and it looks good to me :+1: . Thanks!

[the remark this answers is not in the document]
Comment (zakariaelas) [reply packages/medusa-plugin-typesense/src/services/typesense.js:49]: Correct! As @olivermrbl explained, the filter param does not follow a specific structure and is vendor-specific. Took a look at what you currently have and it looks good to me +1 . Thanks!

**Comment by jasonbosco** [comment]
@olivermrbl I tried to setup a local medusa server and install the plugin following your instructions above and I ran into this error: 

```
$ yarn link "medusa-plugin-typesense"
yarn link v1.22.17
success Using linked package for "medusa-plugin-typesense".
✨  Done in 0.05s.
medusa-testground (master)$ medusa develop
Successfully compiled 0 files with Babel (4ms).
info:    Using fake Redis
✔ Models initialized – 69ms
✔ Plugin models initialized – 3ms
✔ Repositories initialized – 57ms
✔ Database initialized – 705ms
✔ Services initialized – 199ms
✔ Express intialized – 10ms
⠋ Initializing pluginserror:   Services must inherit from BaseService, please check /medusa/packages/medusa-plugin-typesense/services/typesense.js
Error: Services must inherit from BaseService, please check /medusa/packages/medusa-plugin-typesense/services/typesense.js
    at /medusa-testground/node_modules/@medusajs/medusa/dist/loaders/plugins.js:286:51
    at step (/medusa-testground/node_modules/@medusajs/medusa/dist/loaders/plugins.js:33:23)
    at Object.next (/medusa-testground/node_modules/@medusajs/medusa/dist/loaders/plugins.js:14:53)
    at /medusa-testground/node_modules/@medusajs/medusa/dist/loaders/plugins.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (/medusa-testground/node_modules/@medusajs/medusa/dist/loaders/plugins.js:4:12)
    at /medusa-testground/node_modules/@medusajs/medusa/dist/loaders/plugins.js:274:87
    at Array.map (<anonymous>)
⠦ Initializing plugins
```

In typesense.js I'm inheriting from `SearchService` which seems to inherit from `BaseService`... Any pointers on how to fix this?

**Comment by olivermrbl** [comment]
> In typesense.js I'm inheriting from `SearchService` which seems to inherit from `BaseService`... Any pointers on how to fix this?

@jasonbosco Ah sorry for not adding this to the steps. When you are linking locally, the interfaces used in the plugin, Medusa server, and core have to be of the same version.

Doing the following should resolve the issue:
- Navigate to your Medusa server project
- `cd node_modules/medusa-interfaces && yarn link`
- Navigate back to root of your Medusa server project
- `cd node_modules/medusa-plugin-typesense && yarn link medusa-interfaces`
- Navigate back to root of your Medusa server project
- `yarn start`

**Comment by jasonbosco** [comment]
@olivermrbl I get this error now:

```shellsession
medusa-testground (master)$ cd node_modules/medusa-interfaces
medusa-interfaces (feat/typesense)$ yarn link
yarn link v1.22.17
success Registered "medusa-interfaces".
info You can now run `yarn link "medusa-interfaces"` in the projects where you want to use this package and it will be used instead.
✨  Done in 0.05s.
medusa-interfaces (feat/typesense)$ cd ..
node_modules (master)$ cd ..
medusa-testground (master)$ cd node_modules/medusa-plugin-typesense
medusa-plugin-typesense (feat/typesense)$ yarn link medusa-interfaces
yarn link v1.22.17
success Using linked package for "medusa-interfaces".
✨  Done in 0.06s.
medusa-plugin-typesense (feat/typesense)$ cd ..
node_modules (master)$ cd ..
medusa-testground (master)$ yarn start
yarn run v1.22.17
$ medusa develop
Successfully compiled 0 files with Babel (4ms).
medusa start

Start development server.

Options:
  --verbose                Turn on verbose output                                                                                                                                                                                                                                                [boolean] [default: false]
  --no-color, --no-colors  Turn off the color in output                                                                                                                                                                                                                                          [boolean] [default: false]
  --json                   Turn on the JSON logger                                                                                                                                                                                                                                               [boolean] [default: false]
  -H, --host               Set host. Defaults to localhost                                                                                                                                                                                                                                  [string] [default: "localhost"]
  -p, --port               Set port. Defaults to 9000                                                                                                                                                                                                                                            [string] [default: "9000"]
  -h, --help               Show help                                                                                                                                                                                                                                                                              [boolean]
  -v, --version            Show version number                                                                                                                                                                                                                                                                    [boolean]
TypeError: cmd is not a function
    at /Users/jason/Programming/jprojs/typesense/medusa-testground/node_modules/@medusajs/medusa-cli/dist/create-cli.js:284:7
    at /Users/jason/Programming/jprojs/typesense/medusa-testground/node_modules/@medusajs/medusa-cli/dist/create-cli.js:108:24
    at Object.handler (/Users/jason/Programming/jprojs/typesense/medusa-testground/node_modules/@medusajs/medusa-cli/dist/create-cli.js:54:24)
    at Object.runCommand (/Users/jason/Programming/jprojs/typesense/medusa-testground/node_modules/yargs/build/lib/command.js:196:48)
    at Object.parseArgs [as _parseArgs] (/Users/jason/Programming/jprojs/typesense/medusa-testground/node_modules/yargs/build/lib/yargs.js:1043:55)
    at Object.parse (/Users/jason/Programming/jprojs/typesense/medusa-testground/node_modules/yargs/build/lib/yargs.js:584:29)
    at module.exports (/Users/jason/Programming/jprojs/typesense/medusa-testground/node_modules/@medusajs/medusa-cli/dist/create-cli.js:392:6)
    at Object.<anonymous> (/Users/jason/Programming/jprojs/typesense/medusa-testground/node_modules/@medusajs/medusa-cli/dist/index.js:70:27)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
```

**Comment by riqwan** [comment]
Hey, thanks for the PR! Since v2 brought a lot of architectural and API changes on the backend, we will be closing this ticket.

**Pull request closed without merging**